File: Second_Year/quant_marketing/Final/Code/Chebyshev-Approximator.jl
#	Chebyshev-Approximator.R
#	-------------------------------------------------------------------------------------
#	Günter J. Hitsch, April 2014
#	Chebyshev approximation for arbitrary number of dimensions
#   Franco's Version in .jl


#	initializeChebyshevApproximator
#	-------------------------------------------------------------------------------------
#	Inputs:
#	N_dim		... dimensionality of approximation problem
#	N_degree	... degree of Chebyshev polynomial
#	N_nodes		... number of Chebyshev interpolation nodes
#	lower_bound	... lower bounds of rectangle
#	upper_bound	... upper bounds of rectangle
#
#	Output: list "cheb"
#	theta		... Chebyshev coefficients for current approximation
#	X			... matrix of all nodes
#	TX			... regressor matrix of Chebyshev polynomials
#	T2			... pre-computed Chebyshev sum-of-squares


using Kronecker
using Distributions

mutable struct ChebyshevFeatures
    N_dim::Int
    N_degree::Int 
    N_nodes::Int
    lower_bound::Array 
    upper_bound::Array
    theta::Array
    X::Array
    TX::Array
    T2::Array
end


function initializeChebyshevApproximator(N_dim, N_degree, N_nodes, lower_bound, upper_bound)
   
    # Catch some basic input errors
	if N_nodes < N_degree + 1
		print("Error: Number of Chebyshev interpolation nodes (N_nodes) < N_degree + 1!\n")
    end
	
	if sum(lower_bound >= upper_bound) > 0
		print("Error: Rectangle (grid) upper_bound > lower_bound not true in all dimensions!\n")
    end

    # Calculate original [-1, 1] nodes
	nu_nodes = .- cos.(pi .* ( (2 .* Array(1:N_nodes) .- 1) ./ (2 * N_nodes)))

	# Create a table of all N_dim dimensional nodes (N_nodes^N_dim by N_dim matrix)
    list_nodes_nu = ()
    for z=1:N_dim
            list_nodes_nu = (list_nodes_nu..., nu_nodes)
    end
    X_temp = collect(Base.product(list_nodes_nu...))
    X_temp = reshape(X_temp, (length(nu_nodes) ^ N_dim))    
    
    # Placeholder
    X = zeros(length(nu_nodes) ^ N_dim, N_dim)
    for ii in 1:length(nu_nodes) ^ N_dim
        X[ii,:] = collect(X_temp[ii])'
    end
	X = X[:, end:-1:1] # Flip.

    # Adjust nodes to range of the interval in dimension k
	delta = upper_bound .- lower_bound
    X = 0.5 .* delta' .* (X .+ 1) .+ lower_bound'

    # TODO:::
	# Calculate Chebyshev polynomial values for all nodes (N_nodes by N_degree+1 matrix)
	Tnu = calculateChebyshevPolynomials(N_degree, nu_nodes)
	
	# Calculate TX, the regressor matrix of Chebyshev polynomials
	TX = copy(Tnu)
	if N_dim > 1
		for k = 2:N_dim 
            TX = TX ⊗ Tnu
        end
    end
	
	# Precompute the denominator "sum of squares" terms used in the Chebyshev coefficients formula
	K = (N_degree + 1)^N_dim
    T2 = sum(TX .* TX, dims=1)'

	# Initialize Chebyshev regression coefficients
	theta = zeros(K, 1)
	
	cheb = ChebyshevFeatures(N_dim, 
                                N_degree, 
                                N_nodes,
                                lower_bound, 
                                upper_bound,
                                theta, 
                                X, 
                                TX, 
                                T2)

    return cheb

end


function calculateChebyshevPolynomials(N_degree, x)
	
	Tx = zeros(length(x), N_degree + 1)

	# Recursively calculate polynomials
	Tx[:,1] .= 1
	if N_degree >= 1 
        Tx[:,2] .= x					# If polynomial degree >= 1
    end
	if N_degree >= 2	            # If polynomial degree >= 2
		for k in 2:N_degree
			Tx[:,k+1] .= 2 .* x .* Tx[:,k] .- Tx[:,k-1]
        end
    end

	return Tx
end


function calculateChebyshevCoefficients(f, cheb) 
	
    theta_hat = (cheb.TX' * f(cheb.X)) ./ cheb.T2

    setfield!(cheb, :theta, theta_hat)

	return cheb
end


function evaluateChebyshev(Z, cheb) 
	
	L = size(Z)[1]
    K = cheb.N_dim
		
	# Adjust points in Z to rectangle
    Z = 2 .* (Z .- cheb.lower_bound' ) ./ (cheb.upper_bound .- cheb.lower_bound)' .- 1
	
	# Polynomial values for each row in Z
	TZ = zeros(L, (cheb.N_degree + 1)^cheb.N_dim)

    for i = 1:L 	
		z = calculateChebyshevPolynomials(cheb.N_degree, Z[i,K])
		if K > 1
			for k =  1:K-1 
                z = calculateChebyshevPolynomials(cheb.N_degree, Z[i,K-1 - (k-1)]) ⊗ z
            end
        end
		TZ[i,:] = z
    end
	
	y_hat = TZ * cheb.theta
	
	return y_hat
end

File: Second_Year/quant_marketing/Final/Code/Chebyshev-Approximator.R
#	Chebyshev-Approximator.R
#	-------------------------------------------------------------------------------------
#	Günter J. Hitsch, April 2014
#	Chebyshev approximation for arbitrary number of dimensions



#	initializeChebyshevApproximator
#	-------------------------------------------------------------------------------------
#	Inputs:
#	N_dim		... dimensionality of approximation problem
#	N_degree	... degree of Chebyshev polynomial
#	N_nodes		... number of Chebyshev interpolation nodes
#	lower_bound	... lower bounds of rectangle
#	upper_bound	... upper bounds of rectangle
#
#	Output: list "cheb"
#	theta		... Chebyshev coefficients for current approximation
#	X			... matrix of all nodes
#	TX			... regressor matrix of Chebyshev polynomials
#	T2			... pre-computed Chebyshev sum-of-squares

initializeChebyshevApproximator = function(N_dim, N_degree, N_nodes, lower_bound, upper_bound) {
		
	# Catch some basic input errors
	if (N_nodes < N_degree + 1)
		stop("Number of Chebyshev interpolation nodes (N_nodes) < N_degree + 1!\n")
	
	if (sum(lower_bound >= upper_bound) > 0)
		stop("Rectangle (grid) upper_bound > lower_bound not true in all dimensions!\n")
	
	
	# Calculate original [-1, 1] nodes
	nu_nodes = -cos(pi*((2*(1:N_nodes) - 1)/(2*N_nodes)))
	
	# Create a table of all N_dim dimensional nodes (N_nodes^N_dim by N_dim matrix)
	xi_nodes = list()
	for (k in 1:N_dim) xi_nodes[[k]] = nu_nodes
	X = as.matrix(expand.grid(xi_nodes))
	X = X[, N_dim:1]								# Flip: Represents order more intuitively
	
	# Adjust nodes to range of the interval in dimension k
	delta = upper_bound - lower_bound
	for (k in 1:N_dim)
		X[,k] = 0.5*delta[k]*(X[,k] + 1) + lower_bound[k]

	# Calculate Chebyshev polynomial values for all nodes (N_nodes by N_degree+1 matrix)
	Tnu = calculateChebyshevPolynomials(N_degree, nu_nodes)
	
	# Calculate TX, the regressor matrix of Chebyshev polynomials
	TX = Tnu
	if (N_dim > 1) {
		for (k in 2:N_dim) TX = TX %x% Tnu
	}
	
	# Precompute the denominator "sum of squares" terms used in the Chebyshev coefficients formula
	K = (N_degree + 1)^N_dim
	T2 = matrix(0, nrow = K, ncol = 1)
	for (k in 1:K)
		T2[k] = t(TX[,k]) %*% TX[,k]
	
	# Initialize Chebyshev regression coefficients
	theta = matrix(0, nrow = K, ncol = 1)
	
	cheb = list(N_dim = N_dim, N_degree = N_degree, N_nodes = N_nodes,
				lower_bound = lower_bound, upper_bound = upper_bound,
				theta = theta, X = X, TX = TX, T2 = T2)
		
	return(cheb)
}


#	calculateChebyshevPolynomials
#	-------------------------------------------------------------------------------------
#	Calculates the Chebyshev polynomials up to degree N_degree for each point in the
#	the vector x

calculateChebyshevPolynomials = function(N_degree, x) {
	
	Tx = matrix(0, nrow = length(x), ncol = N_degree + 1)

	# Recursively calculate polynomials
	Tx[,1] = 1
	if (N_degree >= 1) Tx[,2] = x					# If polynomial degree >= 1
	if (N_degree >= 2)	{							# If polynomial degree >= 2
		for (k in 2:N_degree)
			Tx[,k+1] = 2*x*Tx[,k] - Tx[,k-1]
	}

	return(Tx)
}


#	calculateChebyshevCoefficients
#	-------------------------------------------------------------------------------------
#	Approximation of function f.  Uses regression to update the Chebyshev coefficients

calculateChebyshevCoefficients = function(f, cheb) {
	
	cheb$theta = (t(cheb$TX) %*% f(cheb$X)) / cheb$T2
	
	return(cheb)
}


#	evaluateChebyshev
#	-------------------------------------------------------------------------------------
#	Predict approximated function values for each row in Z

evaluateChebyshev = function(Z, cheb) {
	
	L = dim(Z)[1]
	K = dim(Z)[2]
		
	# Adjust points in Z to rectangle
	for (k in 1:K)
		Z[,k] = 2*(Z[,k]-cheb$lower_bound[k])/(cheb$upper_bound[k] - cheb$lower_bound[k]) - 1
	
	# Polynomial values for each row in Z
	TZ = matrix(0, nrow = L, ncol = (cheb$N_degree + 1)^cheb$N_dim)
	for (i in 1:L) {	
		z = calculateChebyshevPolynomials(cheb$N_degree, Z[i,K])
		if (K > 1) {
			for (k in (K-1):1) z = calculateChebyshevPolynomials(cheb$N_degree, Z[i,k]) %x% z
		}
		TZ[i,] = z
	}
	
	y_hat = TZ %*% cheb$theta
	
	return(y_hat)
}

